rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,611 @@
+
+### Out of scope
+
+- Validating a component against all possible permutations of its configuration.
+- Validating functional behavior of a component i.e. a transform that should lowercase all strings
+  actually lowercased all strings, etc.
+- Validating any component types besides sources, transforms, and sinks.
+
+## Pain
+
+Currently, there is a significant amount of pain around knowing that a component emits the
+intended/relevant telemetry for that component, and that the values emitted are correct. As
+developers providing support, and operators actually running Vector, the ability to build dashboards
+and alerting that work consistently for different component types is of the utmost importance.
+
+The current process of validating components is inherently manually: we can write a unit test to
+check for compliance, but if we forget to write that unit test, or don't update all unit tests when
+we want to change the behavior for, say, all sources... then our compliance checking can quickly
+become out-of-sync with reality.
+
+This goes double for commuinity contributions: while the Vector authors might have learned over time
+to make sure they're emitting the right metrics, and so on, it's essentially impossible for a
+first-time contributor to know to do so unless there is some mechanism in place to tell them to do
+so.
+
+Humans are also inherently fallible: it is easy to look at data and misread a number, to think
+something is there when it isn't. This has occurred multiple times -- as evidenced by the multiple
+issues linked in the **Context** section -- during work to validate components, despite multiple
+engineers participating in the work, each of them giving it a fresh look.
+
+## Proposal
+
+### User Experience
+
+Instead of having to manually add unit tests to an individual component, components would be
+automatically tested if they were usable. Being "usable", in this case, means that the component is
+present in a given Vector binary: if we're building Vector in a way that includes, say, component X,
+its mere inclusion in the binary will ensure that component X gets tested. It should effectively be
+impossible to include a component in a build in a way that doesn't surface it to be tested.
+
+These components would be validated for compliance by being run automatically in a tailored
+validation harness called the "validation runner." This runner would construct an isolated Vector
+topology that ran the component, and provided all of the necessary inputs/outputs for it to be a
+valid topology. By controlling the input events sent to the topology, and collecting the output
+events and telemetry, we would ensure thhat the component could be tested in isolation and in a
+scenario that looked more like Vector running in a customer's environment than a simple unit test.
+
+This runner would run within a unit test itself, being fed a list of components to validate, as well
```

Comment:
If we're going to run the topology in-process, it may make sense to leverage some of the [existing mock sources/sinks](https://github.com/vectordotdev/vector/blob/ea17c1f2c95c3b6b950f20079f1ec11b6efe03b7/src/test_util/mock/sources/basic.rs#L94) that we already use for the topology tests.

Reply:
The proposal uses the `vector` source and sink to fill in the topology as necessary.

My reasoning for this is that it lets us trivially, if desired, migrate this validation runner to a true "create a configuration and launch Vector with it" approach. It does mean we have to create a gRPC client/server depending on which components we're mocking out to create a valid, connected topology... and that's not as simple or lightweight, relatively speaking, as the true mock components used for testing.

I also wanted to use "real" sources to make sure that mock components couldn't possibly do anything "wrong" with the data that might jeopardize the validation run, or potentially create hard-to-diagnose problems, like false acknowledgements or subtly modifying data or something like that.